[PT-D][Composability][1/N] Upstream NamedOptimizer from TorchRec (KeyedOptimizer in TR)

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89480

In pytorch, the optim state_dict will always use number to index optimizer state_dict for parameters.

Now composability workstream need a FQN based way to index optimizer state_dict for parameters..

For example, SGD optimizer might have something in its `state_dict` like:

```
{'state':
  {0:
    {'momentum_buffer': tensor(...)},
  {1:
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': [0, 1, 2, 3, 4, 5, 6, 7]}]
}
```

And in NamedOptimizer we want the `state_dict` can be:

```
{'state':
  {'net1.0.weight':
    {'momentum_buffer': tensor(...)},
  {'net1.0.bias':
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': ['net1.0.weight', 'net1.0.bias', 'net2.0.weight', 'net2.0.bias', 'net3.weight', 'net3.bias', 'net4.1.weight', 'net4.1.bias']}]
}
```

We also want to support load_state_dict to enable optim `state_dict` override for NameOptimizer.

For the next couple PR/diffs, we also need to:
1. To make `NamedOptimizer` working with FSDP (like registering a hook for model wrapped with FSDP) and other PTD/PT components.
2. Make `NamedOptimizer` works well with apply_optim_in_backward
3. Upstream also `CombinedOptimizer`.

Differential Revision: [D41432088](https://our.internmc.facebook.com/intern/diff/D41432088/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41432088/)!

Commit:
[PT-D][Composability][1/N] Upstream NamedOptimizer from TorchRec (KeyedOptimizer in TR)

In pytorch, the optim state_dict will always use number to index optimizer state_dict for parameters.

Now composability workstream need a FQN based way to index optimizer state_dict for parameters..

For example, SGD optimizer might have something in its `state_dict` like:

```
{'state':
  {0:
    {'momentum_buffer': tensor(...)},
  {1:
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': [0, 1, 2, 3, 4, 5, 6, 7]}]
}
```

And in NamedOptimizer we want the `state_dict` can be:

```
{'state':
  {'net1.0.weight':
    {'momentum_buffer': tensor(...)},
  {'net1.0.bias':
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': ['net1.0.weight', 'net1.0.bias', 'net2.0.weight', 'net2.0.bias', 'net3.weight', 'net3.bias', 'net4.1.weight', 'net4.1.bias']}]
}
```

We also want to support load_state_dict to enable optim `state_dict` override for NameOptimizer.

For the next couple PR/diffs, we also need to:
1. To make `NamedOptimizer` working with FSDP (like registering a hook for model wrapped with FSDP) and other PTD/PT components.
2. Make `NamedOptimizer` works well with apply_optim_in_backward
3. Upstream also `CombinedOptimizer`.

Differential Revision: [D41432088](https://our.internmc.facebook.com/intern/diff/D41432088/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41432088/)!

[ghstack-poisoned]

Commit:
Update on "[PT-D][Composability][1/N] Upstream NamedOptimizer from TorchRec (KeyedOptimizer in TR)"


In pytorch, the optim state_dict will always use number to index optimizer state_dict for parameters.

Now composability workstream need a FQN based way to index optimizer state_dict for parameters..

For example, SGD optimizer might have something in its `state_dict` like:

```
{'state':
  {0:
    {'momentum_buffer': tensor(...)},
  {1:
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': [0, 1, 2, 3, 4, 5, 6, 7]}]
}
```

And in NamedOptimizer we want the `state_dict` can be:

```
{'state':
  {'net1.0.weight':
    {'momentum_buffer': tensor(...)},
  {'net1.0.bias':
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': ['net1.0.weight', 'net1.0.bias', 'net2.0.weight', 'net2.0.bias', 'net3.weight', 'net3.bias', 'net4.1.weight', 'net4.1.bias']}]
}
```

We also want to support load_state_dict to enable optim `state_dict` override for NameOptimizer.

For the next couple PR/diffs, we also need to:
1. To make `NamedOptimizer` working with FSDP (like registering a hook for model wrapped with FSDP) and other PTD/PT components.
2. Make `NamedOptimizer` works well with apply_optim_in_backward
3. Upstream also `CombinedOptimizer`.

Differential Revision: [D41432088](https://our.internmc.facebook.com/intern/diff/D41432088/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41432088/)!

[ghstack-poisoned]

Commit:
Update on "[PT-D][Composability][1/N] Upstream NamedOptimizer from TorchRec (KeyedOptimizer in TR)"


In pytorch, the optim state_dict will always use number to index optimizer state_dict for parameters.

Now composability workstream need a FQN based way to index optimizer state_dict for parameters..

For example, SGD optimizer might have something in its `state_dict` like:

```
{'state':
  {0:
    {'momentum_buffer': tensor(...)},
  {1:
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': [0, 1, 2, 3, 4, 5, 6, 7]}]
}
```

And in NamedOptimizer we want the `state_dict` can be:

```
{'state':
  {'net1.0.weight':
    {'momentum_buffer': tensor(...)},
  {'net1.0.bias':
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': ['net1.0.weight', 'net1.0.bias', 'net2.0.weight', 'net2.0.bias', 'net3.weight', 'net3.bias', 'net4.1.weight', 'net4.1.bias']}]
}
```

We also want to support load_state_dict to enable optim `state_dict` override for NameOptimizer.

For the next couple PR/diffs, we also need to:
1. To make `NamedOptimizer` working with FSDP (like registering a hook for model wrapped with FSDP) and other PTD/PT components.
2. Make `NamedOptimizer` works well with apply_optim_in_backward
3. Upstream also `CombinedOptimizer`.

Differential Revision: [D41432088](https://our.internmc.facebook.com/intern/diff/D41432088/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41432088/)!

[ghstack-poisoned]

wanchaol commented:
Could we explain the exact motivation of upstreaming named optimizer? Last time synced with TorchRec folks the primary reason for things like NamedOptimizer exist is because their sharder is not SPMD style, so they have to keep the names, but since now their sharding are mostly SPMD, I think keeping qualified names for params and states is not needed anymore?

awgu commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,237 @@
+                storage_reader=fs_storage_loader,
+            )
+
+            optim_state_dict.load_state_dict(optim_state_dict)
```
Typo?
```suggestion
            optimizer.load_state_dict(optim_state_dict)
```

fduwjj replied:
opps, thanks!

awgu commented on torch/distributed/optim/named_optimizer.py:
````diff
@@ -0,0 +1,237 @@
+            ...
+        ```
+        Args:
+            state_dict (Dict[str, Any]) : A state_dict to load to. Note that this
````
I wanted to double check this `state_dict` arg description. What does it mean to load _to_ a state dict? Are we modifying `state_dict` in `load_state_dict()`?

fduwjj replied:
yes. Let me make it more clear.

awgu commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,237 @@
+            state_dict (Dict[str, Any]) : A state_dict to load to. Note that this
+                state dict will updated in places.
+        """
+        new_state_dict = self._optimizer.state_dict()
```
I am trying to get a sense of the design principle here.

For FSDP, we similarly need to apply a `FlatParameter`-aware transformation to the state dict before loading. Does it make sense for `NamedOptimizer.load_state_dict()` to provide a native `ShardedTensor`-aware transformation?

I also want to be wary of re-implementing core loading logic. Would optimizer state dict pre- and post-hooks help here?

fduwjj replied:
For FSDP, we definitely need this transformation in general as you mentioned because `self._optimizer.state_dict()` returns `FlatParameter` right? But we also want to make this class more generic. So I think we do need a hook here. I guess we already have such hooks in FSDP today right? 

For core logic, would I be correct that we can reuse the API `shard_full_optim_state_dict` and `flatten_sharded_optim_state_dict` in the hook?

fegin replied:
These APIs are specific for FSDP. NamedOptimizer should not be aware of the existence of these APIs. NamedOptimizer should provide APIs to let FSDP register those FSDP-internal APIs.

awgu commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,237 @@
+                )
+            for k in src_group:
+                if k not in new_group:
+                    raise ValueError(f"Group key {k} not found for group {group_key}")
```
Same with these other messages: It would be great if we can make explicit that the issue is the passed-in `state_dict` to `load_state_dict()`.

fduwjj replied:
sure.

awgu commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,237 @@
+            state_dict (Dict[str, Any]) : A state_dict to load to. Note that this
+                state dict will updated in places.
+        """
+        new_state_dict = self._optimizer.state_dict()
```
IIUC, if the user calls `NamedOptimizer.load_state_dict(state_dict)` but the wrapped optimizer's state has not been initialized yet, then this `NamedOptimizer.load_state_dict()` method will error somewhere while trying to load (looks like L162 if `len(new_state) != len(state)`).

Does this match the behavior of `torch.optim.Optimizer.load_state_dict()`?

fduwjj replied:
That's a good point, `torch.optim.Optimizer.load_state_dict()` does allow this behavior but keyedOptimizer in TorchRec does not. I don't know what we want from NamedOptimizer for this behavior and I agree we need to handle this case. Let me add a TODO here. This PR it too big, I don't want to include it here.

fegin replied:
I belive TorchRec optimizer provides an API to eagerly initialize the optimizer state.

awgu reviewed (commented):
I tried to make an initial higher-level pass.

fduwjj commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,237 @@
+
+    # pyre-ignore [2]
+    def add_param_group(self, param_group: Any) -> None:
+        raise NotImplementedError()
```
Oh I just follow TorchRec's code here.
https://github.com/pytorch/torchrec/blob/main/torchrec/optim/keyed.py#L204

We can definitely add implementations here if we want.

fduwjj commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,237 @@
+                    assert isinstance(src_state_val, torch.Tensor)
+                    state_val.detach().copy_(src_state_val)
+                else:
+                    new_state[idx][state_key] = deepcopy(src_state_val)
```
Mostly, ShardedTensor and other non-Parameter, non-Buffer objects? I learned this from https://github.com/pytorch/pytorch/issues/62094, but it's for model `state_dict`. Maybe we can ask TorchRec folks why they need this in the first beginning?

awgu replied:
Yes, we can ask TorchRec folks mainly for our understanding (not too high priority).

fduwjj commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,237 @@
+                    ret_group[k] = deepcopy(v)
+            ret_groups.append(ret_group)
+
+        ret: Dict[str, object] = {"state": ret_state}
```
Opps..

fduwjj commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,237 @@
+
+    def __init__(
+        self,
+        named_parameters: Mapping[str, Union[torch.Tensor, ShardedTensor]],
```
I am not aware of any discussions around constructor. Actually this is a great question and that's why I want to send out this PR out as fast as possible. I make the signature like this based my understanding of docs reading and TorchRec code reading. And yes, TorchRec's implementation holds that assumption as well. Also use either `parameters` or `param_groups` seems to be a legit way to initialize a normal optimizer. But I am open to any suggestions or comments.

fduwjj commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,237 @@
+                    for shard, src_shard in zip(
+                        state_val.local_shards(), src_state_val.local_shards()
+                    ):
+                        shard.tensor.detach().copy_(src_shard.tensor)
```
Just a n00b question, is there any difference in terms of perf?

fegin replied:
I don't think you have to worry about the performance. This API is not in the critical path. Code simplicity trumps the minor performance win.

Commit:
Update on "[PT-D][Composability][1/N] Upstream NamedOptimizer from TorchRec (KeyedOptimizer in TR)"


In pytorch, the optim state_dict will always use number to index optimizer state_dict for parameters.

Now composability workstream need a FQN based way to index optimizer state_dict for parameters..

For example, SGD optimizer might have something in its `state_dict` like:

```
{'state':
  {0:
    {'momentum_buffer': tensor(...)},
  {1:
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': [0, 1, 2, 3, 4, 5, 6, 7]}]
}
```

And in NamedOptimizer we want the `state_dict` can be:

```
{'state':
  {'net1.0.weight':
    {'momentum_buffer': tensor(...)},
  {'net1.0.bias':
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': ['net1.0.weight', 'net1.0.bias', 'net2.0.weight', 'net2.0.bias', 'net3.weight', 'net3.bias', 'net4.1.weight', 'net4.1.bias']}]
}
```

We also want to support load_state_dict to enable optim `state_dict` override for NameOptimizer.

For the next couple PR/diffs, we also need to:
1. To make `NamedOptimizer` working with FSDP (like registering a hook for model wrapped with FSDP) and other PTD/PT components.
2. Make `NamedOptimizer` works well with apply_optim_in_backward
3. Upstream also `CombinedOptimizer`.

Differential Revision: [D41432088](https://our.internmc.facebook.com/intern/diff/D41432088/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41432088/)!

[ghstack-poisoned]

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41432088).

Commit:
Update on "[PT-D][Composability][1/N] Upstream NamedOptimizer from TorchRec (KeyedOptimizer in TR)"


In pytorch, the optim state_dict will always use number to index optimizer state_dict for parameters.

Now composability workstream need a FQN based way to index optimizer state_dict for parameters..

For example, SGD optimizer might have something in its `state_dict` like:

```
{'state':
  {0:
    {'momentum_buffer': tensor(...)},
  {1:
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': [0, 1, 2, 3, 4, 5, 6, 7]}]
}
```

And in NamedOptimizer we want the `state_dict` can be:

```
{'state':
  {'net1.0.weight':
    {'momentum_buffer': tensor(...)},
  {'net1.0.bias':
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': ['net1.0.weight', 'net1.0.bias', 'net2.0.weight', 'net2.0.bias', 'net3.weight', 'net3.bias', 'net4.1.weight', 'net4.1.bias']}]
}
```

We also want to support load_state_dict to enable optim `state_dict` override for NameOptimizer.

For the next couple PR/diffs, we also need to:
1. To make `NamedOptimizer` working with FSDP (like registering a hook for model wrapped with FSDP) and other PTD/PT components.
2. Make `NamedOptimizer` works well with apply_optim_in_backward
3. Upstream also `CombinedOptimizer`.

Differential Revision: [D41432088](https://our.internmc.facebook.com/intern/diff/D41432088/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41432088/)!

[ghstack-poisoned]

fegin commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+                Note that this state dict update is performed in places.
+        """
+        # TODO: Need to handle the case when self._optimizer has not been initialized.
+        new_state_dict = self._optimizer.state_dict()
```
Can we add a comment to rationale why getting the existing `state_dict` is required?

rohan-varma commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+            optimizer = NamedOptimizer(my_model.named_parameters(), Adagrad)
+            ...
+
+            optim_state_dict = optimizer.state_dict()
```
nit: do we need this saved code sample in `load_state_dict` portion?

rohan-varma commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+    def load_state_dict(self, state_dict: Mapping[str, Any]) -> None:
+        """
+        This public function defines the default behavior to load a state_dict
+        for named optimizer.
```
nit: `NamedOptimizer`

rohan-varma commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+                Note that this state dict update is performed in places.
+        """
+        # TODO: Need to handle the case when self._optimizer has not been initialized.
+        new_state_dict = self._optimizer.state_dict()
```
I think this is essentially the same as @awgu 's comment - but have we documented what we expect in terms of lazy state initialization of the wrapped optimizer, and how this differs today from torch.optim.Optimizer? It seems like it would be useful to include this in docstring.

fduwjj replied:
Sure.

rohan-varma reviewed (approved):
LGTM overall, stamping to unblock, but PTAL at all of the comments.

Thanks for this amazing effort!

fduwjj commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+logger = logging.getLogger(__name__)
+
+
+class NamedOptimizer(optim.Optimizer):
```
I see. Sure. Let's first make it experimental. Thanks for calling that out.

fduwjj commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+            *args,
+            **kwargs,
+        )
+        if param_groups is None:
```
Let me add a TODO here and I will address in a separate PR.

fduwjj commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+            optimizer = NamedOptimizer(my_model.named_parameters(), Adagrad)
+            ...
+
+            optim_state_dict = optimizer.state_dict()
```
To me this is one of the use case for NamedOptimizer? Maybe I can make it more generic.

awgu commented on test/distributed/optim/test_named_optimizer.py:
```diff
@@ -0,0 +1,189 @@
+                {"params": m_dup.net4.parameters(), "lr": 1e-5},
+            ],
+        )
+        for i in range(10):
```
nit: Do we really need more than 1 iteration to create optimizer state?

Maybe change this to just `2` like you do in `test_load_state_dict` to save some computation? (not that the computation is the bottleneck though)

fduwjj replied:
Oh right.. I previously thought it might change after couple iterations. But I agree that 2 is enough here.

awgu commented on test/distributed/optim/test_named_optimizer.py:
```diff
@@ -0,0 +1,189 @@
+
+
+class NamedOptimizerTest(unittest.TestCase):
+    def _compare_state_dict_group(self, group, named_group, compare_equal=True):
```
nit: As a personal recommendation, I prefer to avoid default arguments when possible for internal methods. The only tradeoff is verbosity, but that verbosity can be helpful when reading.

For example below in `test_load_state_dict`, if you pass `False` as the value for `compare_equal` vs. if you pass nothing, it is not immediately obvious what that means. A developer would have to take the additional step to (re-)read this method.

In contrast, imagine if you wrote:
```
self._compare_state_dict_groups(group1, group2, assert_equal=False)
...
self._compare_state_dict_groups(group1, group2, assert_equal=True)
```
(I felt that `assert_equal` was clearer than `compare_equal`, but that is just personal choice.)
In this case, it is more clear what the method call is doing at the callsite without requiring an additional read of the `_compare_state_dict_group` implementation. I can tell the first call is checking `group1` and `group2` are _not equal_ and that the second call is checking that they are _equal_.

fduwjj replied:
Oh, I see. That makes sense. Will do.

awgu commented on torch/distributed/optim/__init__.py:
```diff
@@ -18,6 +18,7 @@
 from .functional_adamax import _FunctionalAdamax
 from .utils import as_functional_optim
 from .apply_optimizer_in_backward import _apply_optimizer_in_backward
+from .named_optimizer import NamedOptimizer
```
Before this PR and can be done in a follow-up: Should we sort these imports?

fduwjj replied:
Sure, let me run the formatter

awgu commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+
+class NamedOptimizer(optim.Optimizer):
+    """
+    NamedOptimizer takes a dict of parameters and exposes state_dict by parameter key.
```
nit: As a heads up, if we want this class to be public in the future, we need to use double backtick (\'\') to surround text that should be rendered as code (e.g. \'\'state_dict\'\' here).

fduwjj replied:
Ah I see. Thanks!

awgu commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+    NamedOptimizer takes a dict of parameters and exposes state_dict by parameter key.
+    We replace the original key (number) in an optim to the FQN string. User can
+    initialize the optim as they initialize a PyTorch optim, the only difference is
+    that they also need to passed in the FQN of each parameters.
```
nit:
```suggestion
    that they also need to pass in the FQN of each parameters.
```

awgu commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+
+    Args:
+        named_parameters (Mapping[str, Union[torch.Tensor, ShardedTensor]]):
+            Parameters of the module and its FQN.
```
nit: Maybe we want to make explicit in the description what is the directionality of the mapping (even though the type annotation does show that)?
```suggestion
            Mapping from FQN to parameter.
```

awgu commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+        optimizer_class (optim.Optimizer):
+            the class of optimizer to instantiate.
+        param_groups (Collection[Mapping[str, Any]]):
+            param_groups to pass to optimizer if specified.
```
Let us specify in this description that the keys in the inner `Mapping` are the FQNs? (In other words, from this description, it is not clear how to specify the `param_groups`.)

fduwjj replied:
Sure.

awgu commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+        >>> # Call state_dict for the named optimizer returns a FQN state_dict.
+        >>> named_optim.state_dict()
+
+    __ TODO: Add tutorial for NamedOptimizer.
```
Curious: What is the `__` for?

fduwjj replied:
NVM, I think this should be removed.. No need to have it here.

awgu commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+        *args,
+        **kwargs,
+    ) -> None:
+        torch._C._log_api_usage_once("torch.distributed.optim.NamedOptimizer")
```
nit: Remember to change this line if you make `NamedOptimizer` private.
```suggestion
        torch._C._log_api_usage_once("torch.distributed.optim._NamedOptimizer")
```

awgu commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+        >>> named_optim.state_dict()
+
+    __ TODO: Add tutorial for NamedOptimizer.
+    """
```
When we make `NamedOptimizer` public, should we add documentation in the docstring for the public attributes like `self.param_groups` and `self.named_parameters`?

fduwjj replied:
Sure. Let me add a TODO here.

awgu commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+            **kwargs,
+        )
+        if param_groups is None:
+            self.param_keys_order = list(self.named_parameters.keys())
```
Hmm. I am not sure about the naming here (`param_keys_order`). This seems like the ordered parameter keys, which are the FQNs right?

Maybe we rename this to `ordered_param_names` or `ordered_param_keys`? Does this need to be public?

awgu commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+        """
+        This public function defines the default behavior to load a state_dict
+        for named optimizer.
+
```
For a follow-up: We should make sure this sample code renders as expected. The syntax looks a bit different than other examples I have seen.

fduwjj replied:
Sure, where do you check the render result?

awgu replied:
I think you have to add `NamedOptimizer` to the `distributed.optim.rst` file, but I think the class needs to be public (meaning that if you change it to `_NamedOptimizer`, then it will not render).

https://github.com/pytorch/pytorch/blob/master/docs/source/distributed.optim.rst

Feel free to refer to `fsdp.rst` as an example:
https://github.com/pytorch/pytorch/blob/master/docs/source/fsdp.rst

fduwjj replied:
I see. Thanks. They did mentioned it in https://github.com/pytorch/pytorch/blob/master/docs/source/community/contribution_guide.rst.

awgu commented on torch/distributed/optim/named_optimizer.py:
````diff
@@ -0,0 +1,244 @@
+        ```
+        Args:
+            state_dict (Dict[str, Any]) : A ``state_dict`` to load into the optimizer.
+                Note that this state dict update is performed in places.
````
```suggestion
                Note that this state dict update is performed in place.
```

awgu commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,244 @@
+    """
+    Concatenate all param keys as a unique indentifier for one param group.
+    """
+
```
nit: Extraneous newline?

Commit:
Update on "[PT-D][Composability][1/N] Upstream NamedOptimizer from TorchRec (KeyedOptimizer in TR)"


In pytorch, the optim state_dict will always use number to index optimizer state_dict for parameters.

Now composability workstream need a FQN based way to index optimizer state_dict for parameters..

For example, SGD optimizer might have something in its `state_dict` like:

```
{'state':
  {0:
    {'momentum_buffer': tensor(...)},
  {1:
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': [0, 1, 2, 3, 4, 5, 6, 7]}]
}
```

And in NamedOptimizer we want the `state_dict` can be:

```
{'state':
  {'net1.0.weight':
    {'momentum_buffer': tensor(...)},
  {'net1.0.bias':
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': ['net1.0.weight', 'net1.0.bias', 'net2.0.weight', 'net2.0.bias', 'net3.weight', 'net3.bias', 'net4.1.weight', 'net4.1.bias']}]
}
```

We also want to support load_state_dict to enable optim `state_dict` override for NameOptimizer.

For the next couple PR/diffs, we also need to:
1. To make `NamedOptimizer` working with FSDP (like registering a hook for model wrapped with FSDP) and other PTD/PT components.
2. Make `NamedOptimizer` works well with apply_optim_in_backward
3. Upstream also `CombinedOptimizer`.

Differential Revision: [D41432088](https://our.internmc.facebook.com/intern/diff/D41432088/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41432088/)!

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

Commit:
Update on "[PT-D][Composability][1/N] Upstream NamedOptimizer from TorchRec (KeyedOptimizer in TR)"


In pytorch, the optim state_dict will always use number to index optimizer state_dict for parameters.

Now composability workstream need a FQN based way to index optimizer state_dict for parameters..

For example, SGD optimizer might have something in its `state_dict` like:

```
{'state':
  {0:
    {'momentum_buffer': tensor(...)},
  {1:
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': [0, 1, 2, 3, 4, 5, 6, 7]}]
}
```

And in NamedOptimizer we want the `state_dict` can be:

```
{'state':
  {'net1.0.weight':
    {'momentum_buffer': tensor(...)},
  {'net1.0.bias':
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': ['net1.0.weight', 'net1.0.bias', 'net2.0.weight', 'net2.0.bias', 'net3.weight', 'net3.bias', 'net4.1.weight', 'net4.1.bias']}]
}
```

We also want to support load_state_dict to enable optim `state_dict` override for NameOptimizer.

For the next couple PR/diffs, we also need to:
1. To make `NamedOptimizer` working with FSDP (like registering a hook for model wrapped with FSDP) and other PTD/PT components.
2. Make `NamedOptimizer` works well with apply_optim_in_backward
3. Upstream also `CombinedOptimizer`.

Differential Revision: [D41432088](https://our.internmc.facebook.com/intern/diff/D41432088/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41432088/)!

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

Commit:
Update on "[PT-D][Composability][1/N] Upstream NamedOptimizer from TorchRec (KeyedOptimizer in TR)"


In pytorch, the optim state_dict will always use number to index optimizer state_dict for parameters.

Now composability workstream need a FQN based way to index optimizer state_dict for parameters..

For example, SGD optimizer might have something in its `state_dict` like:

```
{'state':
  {0:
    {'momentum_buffer': tensor(...)},
  {1:
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': [0, 1, 2, 3, 4, 5, 6, 7]}]
}
```

And in NamedOptimizer we want the `state_dict` can be:

```
{'state':
  {'net1.0.weight':
    {'momentum_buffer': tensor(...)},
  {'net1.0.bias':
    {'momentum_buffer': tensor(...)},
  ...
}
'param_groups':
    [{'lr': 0.001, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0, 'nesterov': False, 'maximize': False, 'foreach': None, 'differentiable': False, 'params': ['net1.0.weight', 'net1.0.bias', 'net2.0.weight', 'net2.0.bias', 'net3.weight', 'net3.bias', 'net4.1.weight', 'net4.1.bias']}]
}
```

We also want to support load_state_dict to enable optim `state_dict` override for NameOptimizer.

For the next couple PR/diffs, we also need to:
1. To make `NamedOptimizer` working with FSDP (like registering a hook for model wrapped with FSDP) and other PTD/PT components.
2. Make `NamedOptimizer` works well with apply_optim_in_backward
3. Upstream also `CombinedOptimizer`.

Differential Revision: [D41432088](https://our.internmc.facebook.com/intern/diff/D41432088/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41432088/)!

[ghstack-poisoned]

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41432088).

fduwjj commented:
@pytorchbot merge -f "All failure are related to test_correct_module_names (__main__.TestPublicBindings) Not related to this PR. Also split internal changes into a separate diff."

awgu commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -0,0 +1,258 @@
+from torch.distributed._shard.sharded_tensor import ShardedTensor
+
+
+__all__ = ["_NamedOptimizer"]
```
I am actually not sure if we should include `"_NamedOptimizer"` in `__all__` if we have it as private (with the preceding `_`).

kit1980 replied:
I believe it's the reason for test_correct_module_names failures.
Reverting this PR.

fduwjj replied:
Thanks... Sorry, I thought it was not caused by this pr... Should have read more carefully about the err msg..

kit1980 commented:
@pytorchbot revert -m "Broke test_correct_module_names" -c ingnoredsignal

kit1980 commented:
@pytorchbot revert -m "Broke test_correct_module_names" -c ignoredsignal
